File: dynamo_ts/src/orders/createOrderRequest.ts
import { SNS } from 'aws-sdk';
import { HttpRequest, HttpResponse } from '../http'
import { v4 as uuidv4 } from 'uuid';
import { CreateOrderRequest, validateOrderRequest } from './forms';
// Create SNS service client
const sns = new SNS();

export async function handler(event: HttpRequest): Promise<HttpResponse> {
  const form: CreateOrderRequest = JSON.parse(event.body)
  const validationErr = validateOrderRequest(form)

  if (validationErr) {
    return {
      statusCode: 400,
      headers: {
        'Access-Control-Allow-Origin': '*',
        'Access-Control-Allow-Credentials': true,
      },
      body: validationErr
    }
  }

  const userId = event.pathParameters['userId']
  // Setup the sendMessage parameter object
  const params: SNS.PublishInput = {
    Message: JSON.stringify({
      userId: userId,
      orderNo: uuidv4(),
      name: form.name,
      email: form.email,
      street: form.street,
      housenr: form.housenr,
      city: form.city,
      postcode: form.postcode,
      products: form.products
    }),
    TopicArn: process.env.ORDER_TOPIC
  }

  try {
    const data = await sns.publish(params).promise()
    console.log("Successfully added message", data.MessageId);
  } catch (err) {
    console.error("Error", err);
    return {
      statusCode: 500,
      headers: {
        'Access-Control-Allow-Origin': '*',
        'Access-Control-Allow-Credentials': true,
      },
      body: JSON.stringify(err)
    }
  }

  return {
    statusCode: 200,
    headers: {
      'Access-Control-Allow-Origin': '*',
      'Access-Control-Allow-Credentials': true,
    },
    body: 'Order abgelegt, status pending'
  }
}

File: dynamo_go/lambda/orders/createOrderRequest/createOrderRequest.go
package main

import (
	"encoding/json"
	"github.com/aws/aws-lambda-go/events"
	"github.com/aws/aws-lambda-go/lambda"
	"github.com/aws/aws-sdk-go/aws"
	"github.com/aws/aws-sdk-go/aws/session"
	"github.com/aws/aws-sdk-go/service/sns"
	uuid "github.com/nu7hatch/gouuid"
	"os"
	"webshop/internal/apigw"
	"webshop/internal/forms"
)

func Handler(event events.APIGatewayProxyRequest) (*events.APIGatewayProxyResponse, error) {

	var createForm forms.OrderRequest
	if err := json.Unmarshal([]byte(event.Body), &createForm); err != nil {
		return apigw.CreateResponse(500, err.Error()), err
	}

	newSns := sns.New(session.New())

	createForm.UserId = aws.String(event.PathParameters["userId"])

	id, err := uuid.NewV4()
	if err != nil {
		return apigw.CreateResponse(500, err.Error()), err
	}
	createForm.OrderNo = aws.String(id.String())

	jsonStr, err := json.Marshal(createForm)
	if err != nil {
		return apigw.CreateResponse(500, err.Error()), err
	}

	_, err = newSns.Publish(&sns.PublishInput{
		Message:  aws.String(string(jsonStr)),
		TopicArn: aws.String(os.Getenv("ORDER_TOPIC")),
	})
	if err != nil {
		return apigw.CreateResponse(500, err.Error()), err
	}

	return apigw.CreateResponse(200, "Order abgelegt, Status pending"), err
}

func main() {
	lambda.Start(Handler)
}
